drivers/common/virtio/virtio.c

```diff
@@ -0,0 +1,443 @@
+	}
+	irq_set->argsz = sizeof(struct vfio_irq_set) + sizeof(int);
+	irq_set->count = 1;
+	irq_set->flags = VFIO_IRQ_SET_DATA_EVENTFD |
+			 VFIO_IRQ_SET_ACTION_TRIGGER;
+	irq_set->index = VFIO_PCI_MSIX_IRQ_INDEX;
+	irq_set->start = vec;
+	*(int *)&irq_set->data = fd;
+
+	ret = ioctl(vpdev->vfio_dev_fd, VFIO_DEVICE_SET_IRQS, irq_set);
+	rte_free(irq_set);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Dev %s enabling MSI-X: %s", VP_DEV_NAME(vpdev), strerror(errno));
+		return ret;
+	}
+
+	VIRTIO_OPS(hw)->intr_detect(hw);
+	if(vpdev->msix_status != VIRTIO_MSIX_ENABLED) {
+		PMD_INIT_LOG(ERR, "Dev %s MSI-X not enabled,status: %d", VP_DEV_NAME(vpdev) , vpdev->msix_status);
+		return -EINVAL;
+	}
+
+	if (vec == 0) {
+		if (VIRTIO_OPS(hw)->set_config_irq(hw, 0) ==
+				VIRTIO_MSI_NO_VECTOR) {
+			PMD_INIT_LOG(ERR, "Failed to set config vector");
+			return -EINVAL;
+		}
+	} else if (VIRTIO_OPS(hw)->set_queue_irq(hw, hw->vqs[vec-1], vec) ==
+					 VIRTIO_MSI_NO_VECTOR) {
+		PMD_INIT_LOG(ERR, "Failed to set queue %d vector", vec-1);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+int
+virtio_pci_dev_interrupts_disable(struct virtio_pci_dev *vpdev, int vec)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	struct vfio_irq_set *irq_set;
+	int ret;
+
+	irq_set = rte_zmalloc(NULL, (sizeof(struct vfio_irq_set) + sizeof(int)), 0);
+	if (irq_set == NULL) {
+		PMD_INIT_LOG(ERR, "Dev %s malloc fail", VP_DEV_NAME(vpdev));
+		return -ENOMEM;
+	}
+	irq_set->argsz = sizeof(struct vfio_irq_set) + sizeof(int);
+	irq_set->count = 1;
+	irq_set->flags = VFIO_IRQ_SET_DATA_EVENTFD |
+			 VFIO_IRQ_SET_ACTION_TRIGGER;
+	irq_set->index = VFIO_PCI_MSIX_IRQ_INDEX;
+	irq_set->start = vec;
+	*(int *)&irq_set->data = VFIO_FD_INVALID;
+
+	ret = ioctl(vpdev->vfio_dev_fd, VFIO_DEVICE_SET_IRQS, irq_set);
+	rte_free(irq_set);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Dev %s disabling MSI-X: %s", VP_DEV_NAME(vpdev), strerror(errno));
+		return ret;
+	}
+
+	if (vec == 0) {
+		if (VIRTIO_OPS(hw)->set_config_irq(hw, VIRTIO_MSI_NO_VECTOR) !=
+				VIRTIO_MSI_NO_VECTOR) {
+			PMD_INIT_LOG(ERR, "Failed to unset config vector");
+			return -EINVAL;
+		}
+	} else if (VIRTIO_OPS(hw)->set_queue_irq(hw, hw->vqs[vec-1], VIRTIO_MSI_NO_VECTOR) !=
+					 VIRTIO_MSI_NO_VECTOR) {
+		PMD_INIT_LOG(ERR, "Failed to unset queue %d vector", vec-1);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+void
+virtio_pci_dev_set_status(struct virtio_pci_dev *vpdev, uint8_t status)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+
+	if (status != VIRTIO_CONFIG_STATUS_RESET)
+		status |= VIRTIO_OPS(hw)->get_status(hw);
+
+	VIRTIO_OPS(hw)->set_status(hw, status);
+}
+
+uint8_t
+virtio_pci_dev_get_status(struct virtio_pci_dev *vpdev)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+
+	return VIRTIO_OPS(hw)->get_status(hw);
+}
+
+void
+virtio_pci_dev_reset(struct virtio_pci_dev *vpdev)
+{
+	uint32_t retry = 0;
+	struct virtio_hw *hw = &vpdev->hw;
+
+	VIRTIO_OPS(hw)->set_status(hw, VIRTIO_CONFIG_STATUS_RESET);
+	/* Flush status write and wait device ready max 3 seconds. */
```

Comment:
3 seconds is too less for us during dpu reboots.

it should be close to 120 secs.